### PR TITLE
Add mock-ups to ADAM/GATK pipelines (resolves #206)

### LINKED ISSUE #206: Add mock-ups to ADAM/GATK pipelines
@hannes-ucsc and I discussed this after scrum today.

In mock-up mode, the steps of the pipeline should do the following, given a small fastq and associated bam file (<20mb):

BWA alignment: Downloads input fastq and pre-computed mock.bam file. Renames bam file and writes to the appropriate s3 bucket.
ADAM preprocessing: Runs as usual, but using tiny instances. Produces a preprocessed bam file which is then replaced with an empty bam that is written to s3.
GATK preprocessing: Downloads the input BAM file, writes dummy intermediate files to the jobstore, and writes a blank preprocessed bam file with the appropriate name to s3.
Haplotype caller: Downloads the input preprocessed BAM file, and writes a blank vcf file with the appropriate name to s3.

In addition we will add assertions after the completion of each pipeline step that check that all necessary files are present. These checks will run in both mock-up mode and production mode.

@fnothaft what are your thoughts? I can get started on this today if you think it seems reasonable.

### PR DESCRIPTION
**Pull request by almussel**
Resolves #206. Also refactors ```docker_call``` into the common library. Depends on #226 and #236.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Title is illformed.

We don't do periods at the end of commit messages.

To refer to an issue in a commit, the syntax is ` (resolves #123)`

Your commit, @almussel should refer to a ticket number.

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py [hunk 1]
```diff
@@ -486,6 +489,25 @@ def decrease_node_count(job, nodes_to_add, uuid):
     Job.Runner.addToilOptions(args_parser)
     args = args_parser.parse_args()
 
+    if mock_mode():
+        from mock_inputs import mock_inputs
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]:
```
for k,v in mock_inputs.iteritems():
   setattr( args, k, v )
```

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py [hunk 2]
```diff
@@ -136,8 +136,10 @@ def build_parser():
 
     parser = argparse.ArgumentParser()
 
+    in_production_mode = not mock_mode()
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]: I'd prefer inlining this.

```
in_production_mode
not mock_mode()
```

See how the inlined version is actually shorter?

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py [hunk 3]
```diff
@@ -37,8 +37,8 @@
 from toil.job import Job
 
 from toil_scripts.adam_uberscript.automated_scaling import SparkMasterAddress
-from toil_scripts.batch_alignment.bwa_alignment import docker_call
-from toil_scripts.spark_utils.spawn_cluster import start_spark_hdfs_cluster
+from toil_scripts.batch_alignment.bwa_alignment import docker_call, mock_mode
+from toil_scripts.spark_utils.spawn_cluster import *
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: * imports are nuts. PyCharm has a refactoring that lets you expand these.

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py [hunk 4]
```diff
@@ -118,7 +121,10 @@ def call_adam(masterIP, inputs, arguments):
                 tool = "quay.io/ucsc_cgl/adam:962-ehf--6e7085f8cac4b9a927dc9fb06b48007957256b80",
                 docker_parameters = masterIP.docker_parameters(["--net=host"]),
                 tool_parameters = default_params + arguments,
-                sudo = inputs['sudo'])
+                inputs=[],
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: [] and {} should be the default of inputs and outputs, respectively. 

But do read: http://docs.python-guide.org/en/latest/writing/gotchas/

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py [hunk 5]
```diff
@@ -128,21 +134,40 @@ def remove_file(masterIP, filename, sparkOnToil):
...
+    if sparkOnToil:
+        output = check_output(ssh_call +  ['docker', 'ps'])
+        containerID = next(line.split()[0] for line in output.splitlines() if 'apache-hadoop-master' in line)
+        ssh_call += ['docker', 'exec', containerID]
+
+    try:
+        check_call(ssh_call + ['hdfs', 'dfs', '-truncate', '-w', '10', '/' + filename])
+    except:
+        pass
+
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: Feels like this hunk has nothing to do with mock mode. Should it be committed separately?

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 6]
```diff
@@ -45,6 +45,11 @@
 # nodes and the cluster-growing thread adding nodes.
 cluster_size_lock = threading.RLock()
 
+def mock_mode():
+    """
+    Checks whether the ADAM_GATK_MOCK_MODE environment variable is set
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: This is obvious. What does mock mode mean? That's not so obvious.

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 7]
```diff
@@ -124,53 +129,58 @@ def launch_pipeline(params):
...
                             '--mesosMaster $(hostname -i):5050',
                             '--workDir /var/lib/toil',
-                            '--file_size {fs}',
                             '--logInfo']
+        if mock_mode():
+            pipeline_command = ['ADAM_GATK_MOCK_MODE=1'] + \
+                               pipeline_command + \
+                               ['--dir_suffix /mock']
+        else:    
+            pipeline_command += ['--s3_bucket {b}',
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Why do we not need to pass the rest of these args in mock mode?

File: src/toil_scripts/adam_uberscript/adam_uberscript.py [hunk 8]
```diff
@@ -540,6 +550,8 @@ def main():
     parser = argparse.ArgumentParser(description=main.__doc__,
                                      formatter_class=argparse.RawDescriptionHelpFormatter)
     subparsers = parser.add_subparsers(dest='command')
+    
+    in_production_mode = not mock_mode()
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: See above

**Comment by almussel** [comment]
Arrgh sorry I botched the rebase so a bunch of old commits are showing up. I'm fixing it now.

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 9]
```diff
@@ -174,16 +175,24 @@ def return_input_paths(job, work_dir, ids, *args):
 
     return paths.values()
 
+def mock_mode():
```
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Can't we import this from an __init__.py somewher higher up?

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 10]
```diff
@@ -174,16 +175,24 @@ def return_input_paths(job, work_dir, ids, *args):
...
+    """
+    Checks whether the ADAM_GATK_MOCK_MODE environment variable is set
+    """
+    return True if int(os.environ.get('ADAM_GATK_MOCK_MODE', '0')) else False
 
 def docker_call(work_dir,
                 tool_parameters,
                 tool,
+                inputs,
```
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: See my comment above about defaulting.

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 11]
```diff
@@ -193,6 +202,39 @@ def docker_call(work_dir,
     outfile: file           Filehandle that stderr will be passed to
     sudo: bool              If the user wants the docker command executed as sudo
     """
+
+    for filename in inputs:
+        assert(os.path.isfile(os.path.join(work_dir, filename)))
+
+    if mock == None:
```
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: `is None`

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 12]
```diff
@@ -193,6 +202,39 @@ def docker_call(work_dir,
...
+                file_path = os.path.join(work_dir, filename)
+                if not os.path.exists(file_path):
+                    if url.startswith('s3:'):
+                        _log.info("trying to download: " + url + " to " + file_path)
+                        download_from_s3_url(file_path, url)
+                    else:
+                        try:
+                            download_cmd = ['curl', '-fs', '--retry', '5', '--create-dir', url, '-o', file_path]
+                            subprocess.check_call(download_cmd)
+                        except OSError:
```
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]:
```
except OSError as e:
    if e.errno == errno.ENOENT:
        raise RuntimeError(…)
    else:
        raise
```

Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: And `import errno` at the top.

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 13]
```diff
@@ -222,6 +264,10 @@ def docker_call(work_dir,
     except OSError:
         raise RuntimeError('docker not found on system. Install on all nodes.')
 
+    for filename in outputs.keys():
+        if not os.path.isabs(filename):
```
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: path has a function for making a path absolute.

File: src/toil_scripts/gatk_germline/germline.py [hunk 14]
```diff
@@ -317,9 +324,13 @@ def index(job, shared_ids, input_args):
     output_path = os.path.join(work_dir, 'toil.bam.bai')
     # Call: index the normal.bam
     parameters = ['index', 'toil.bam']
+    inputs=['toil.bam']
+    outputs={"toil.bam.bai": None}
```
Comment (hannes-ucsc) [inline src/toil_scripts/gatk_germline/germline.py]: single quotes by default

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py [hunk 15]
```diff
@@ -269,6 +269,36 @@ def docker_call_preprocess(work_dir, tool_parameters, tool, java_opts=None,
     outfile: file           Filehandle that stderr will be passed to
     sudo: bool              If the user wants the docker command executed as sudo
     """
+
+    for filename in inputs:
```
Comment (hannes-ucsc) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: Maybe this is the time to start consolidating all these `docker_call` instances.

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py [hunk 16]
```diff
@@ -594,14 +658,15 @@ def indel_realignment(job, shared_ids, input_args):
                   '-o', 'sample.indel.bam']
 
     docker_call_preprocess(tool='quay.io/ucsc_cgl/gatk:3.5--dba6dae49156168a909c43330350c6161dc7ecc2',
-                           work_dir=work_dir, tool_parameters=parameters,
-                           java_opts='-Xmx%sg' % input_args['memory'], sudo=sudo,
-                           outfiles=['sample.indel.bam', 'sample.indel.bam.bai'])
+                work_dir=work_dir, tool_parameters=parameters,
```
Comment (hannes-ucsc) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: Oddly indented.

**Comment by hannes-ucsc** [comment]
I feel like 4b37b98 conflates mock mode with some other changes. Should it be split up?

And as I said above, I hate those periods at the end in @fnothaft commits and the fact that those commits don't follow the convention. One of you should fix those things via an interactive rebase.

**Comment by hannes-ucsc** [comment]
… don't follow the convention of how we denote ticket references in commit messages.

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py (hunk 3 above)
Comment (almussel) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: Agreed, but that change was due to some problem with how imports work with Toil. @fnothaft said using ```*``` was necessary.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 13 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: That won't work here because any non-absolute path will be a subdirectory of work_dir, not of the current directory.

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py (hunk 15 above)
Comment (almussel) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: I think that's a good idea. Should I do that in this PR or a separate one?

**Comment by almussel** [comment]
Yes, there are a few unrelated changes I made and included here. I will pull them out and PR separately.

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 7 above)
Comment (almussel) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Those are passed in automatically by the pipeline from a dictionary. I did it that way so that running pipeline tests in mock mode without the überscript would be really straightforward.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 9 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: I deliberately put mock_mode next to docker_call for simplicity. I also put an identical function in the überscript because importing between it and the pipelines introduces a bunch of dependencies.

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py (hunk 2 above)
Comment (almussel) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]: Fixed

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py (hunk 4 above)
Comment (almussel) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: Fixed

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py (hunk 5 above)
Comment (almussel) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: I moved this to a separate PR

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 8 above)
Comment (almussel) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Fixed

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 6 above)
Comment (almussel) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: Fixed

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 10 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Fixed

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 11 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Fixed

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 12 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Fixed

File: src/toil_scripts/gatk_germline/germline.py (hunk 14 above)
Comment (almussel) [inline src/toil_scripts/gatk_germline/germline.py]: Fixed

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py (hunk 16 above)
Comment (almussel) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: Fixed

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py (hunk 3 above)
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: I doubt it.

```
hannes-mbpr:x hannes $ cat foo.py
from threading import Event
hannes-mbpr:x hannes $ cat bar.py
from multiprocessing import Event
from foo import *
assert Event.__module__ == 'multiprocessing', Event.__module__
hannes-mbpr:x hannes $ python bar.py 
Traceback (most recent call last):
  File "bar.py", line 3, in <module>
    assert Event.__module__ == 'multiprocessing', Event.__module__
AssertionError: threading
(bd2k:/__me__/)hannes-mbpr:x hannes $ 
```

This has bitten us with those dupes of the download_url function.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 9 above)
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Duplication of code within a single source tree is nuts. We worked so hard to get rid of the dupes and this is sneaking it back in. Also, proximity isn't the same simplicity. 

Putting it in an `__init__.py` should prevent the dependency issue you mention.

Feel free to discuss in person.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 13 above)
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: You are right.

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py (hunk 15 above)
Comment (hannes-ucsc) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: Separate commit would have worked, too.

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]: Note to self: that's taken care of below.

File: src/toil_scripts/adam_uberscript/adam_uberscript.py (hunk 7 above)
Comment (hannes-ucsc) [inline src/toil_scripts/adam_uberscript/adam_uberscript.py]: I'm still leaning towards minimizing the differences between mock and production mode. But I don't understand the full depth of the issue here. Your call.

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py (hunk 15 above)
Comment (almussel) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: I added the ```docker_call``` refactoring to this PR.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 9 above)
Comment (almussel) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: mock_mode has been moved to lib/programs

**Comment by almussel** [comment]
@hannes-ucsc will you review my updates?

**Comment by fnothaft** [comment]
Thanks @almussel!

**Comment by jvivian** [comment]
In docker_call() I see there are now mutable datatypes set as defaults (`parameters`, `inputs`, `outputs) — Isn't that a no-no?

**Comment by hannes-ucsc** [comment]
> no-no

Correct. I had [pointed](https://github.com/BD2KGenomics/toil-scripts/pull/230#discussion_r59642270) this out and thought it was fixed but didn't check again.

**Comment by hannes-ucsc** [comment]
@almussel, I created https://github.com/BD2KGenomics/toil-scripts/issues/247 to track this. Can you submit a follow on PR?

**Comment by almussel** [comment]
Yes, will do.
